x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java

```diff
@@ -0,0 +1,260 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.idp.saml.authn;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.apache.logging.log4j.util.MessageSupplier;
+import org.elasticsearch.ElasticsearchSecurityException;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.core.internal.io.Streams;
+import org.elasticsearch.rest.RestUtils;
+import org.elasticsearch.xpack.idp.action.SamlValidateAuthnRequestResponse;
+import org.elasticsearch.xpack.idp.saml.idp.SamlIdentityProvider;
+import org.elasticsearch.xpack.idp.saml.sp.SamlServiceProvider;
+import org.elasticsearch.xpack.idp.saml.support.SamlUtils;
+import org.opensaml.saml.saml2.core.AuthnRequest;
+import org.opensaml.saml.saml2.core.Issuer;
+import org.opensaml.saml.saml2.core.NameIDPolicy;
+import org.opensaml.security.x509.X509Credential;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.xml.sax.SAXException;
+
+import javax.xml.parsers.DocumentBuilder;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.security.Signature;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.zip.Inflater;
+import java.util.zip.InflaterInputStream;
+
+import static org.opensaml.saml.saml2.core.NameIDType.UNSPECIFIED;
+
+
+/*
+ * Processes a SAML AuthnRequest, validates it and extracts necessary information
+ */
+public class SamlAuthnRequestValidator {
+
+    private final SamlIdentityProvider idp;
+    private final Logger logger = LogManager.getLogger(SamlAuthnRequestValidator.class);
+    private static final String[] XSD_FILES = new String[]{"/org/elasticsearch/xpack/idp/saml/support/saml-schema-protocol-2.0.xsd",
+        "/org/elasticsearch/xpack/idp/saml/support/saml-schema-assertion-2.0.xsd",
+        "/org/elasticsearch/xpack/idp/saml/support/xenc-schema.xsd",
+        "/org/elasticsearch/xpack/idp/saml/support/xmldsig-core-schema.xsd"};
+
+    private static final ThreadLocal<DocumentBuilder> THREAD_LOCAL_DOCUMENT_BUILDER = ThreadLocal.withInitial(() -> {
+        try {
+            return SamlUtils.getHardenedBuilder(XSD_FILES);
+        } catch (Exception e) {
+            throw new ElasticsearchSecurityException("Could not load XSD schema file", e);
+        }
+    });
+
+    public SamlAuthnRequestValidator(SamlIdentityProvider idp) {
+        SamlUtils.initialize();
+        this.idp = idp;
+    }
+
+    public SamlValidateAuthnRequestResponse processQueryString(String queryString) {
```

Comment:
I'm pretty sure this will end up async (needing to lookup SP in index).
Can we structure it to have a listener now, so that it's a bit simpler to change later on?

Reply:
I left this on purpose for the registration API PR but no objections to tackle this here, will do!